### DISCUSSION TITLE
Unintuitive behaviour when atom setter calls itself

### DISCUSSION THREAD

**Discussion by n-p-e** [pmndrs/jotai#2018] [Bug report]
I've noticed that atom setters don't play very nicely with recursion. Here's an example:

```typescript
import { atom, useAtom } from "jotai";

const badAtom = atom(null, (get, set, arg: string) => {
  console.log(`doing something with ${arg}`);
  setTimeout(
    () => /*problem*/ set(badAtom, "next arg"),
    500
  );
});

export const Page: React.FC = () => {
  const [value, doThing] = useAtom(badAtom);
  return <div>
    <button onClick={() => doThing("first arg")}>Do this</button>
    <p>Atom value is: {String(value)}</p>
  </div>
};
export default Page;
```

What I intended to do is have the setter run itself every 500ms. But after the first execution, the recursive call actually sets `badAtom`'s value to be `"next arg"` instead of calling itself again. There seems to be no other way to have the setter recurse, except defining two atoms and have them call each other (mutual recursion).

![image](https://github.com/pmndrs/jotai/assets/29776557/7cdc46eb-d777-4db8-a03b-200bf92e4aaa)

Another problem is this behaviour breaks type safety. If you hover over the `value` in the `Page` component, you can see its inferred type is `null`, same with the atom's initial value. But at runtime it becomes a string.

I think this problem originates from [here](https://github.com/pmndrs/jotai/blob/a9a639f4589c660b1f0723c536bf69d655bd71be/src/vanilla/store.ts#L531) where we special case for when the `set` target being the atom itself. The type cast introduces this problem, and makes `set(thisAtom, ...)` and `set(anotherAtom, ...)` confusingly do two different things (directly set value vs. invoke setter).

My proposed solution is to provide `set.self(newValue)` as a safer alternative to what `set(thisAtom, value)` does now, and constrain `newValue` to be the same type as the atom's value. And for the recursive use case, provide `set.recurse(args)` as well. In Jotai 3.0, it may be worth it to make `set(thisAtom, value)` a runtime error.

**Comment by dai-shi** [comment]
It's a bit counter-intuitive, but internally there's no notion of write-only atoms. It's just a use case of read-write atoms. We don't prefer adding new apis for recursive setters. It would be nice to be warned somehow though.

The canonical solution is to use two atoms:
```js
const actionAtom = atom(null, (get, set, arg: string) => {
  console.log(`doing something with ${arg}`);
});
const timerAtom = atom(null, (get, set) => {
  setInterval(
    () => set(actionAtom, "next arg"),
    500
  );
});
```

It's not only this case, but we generally do things with multiple atoms instead of one.

**Reply by n-p-e** [reply]
Thanks for the suggestion. I understand new APIs should be carefully considered.

Apart from recursion, I think another problem is the confusion between "set value" and "call write function". Currently, `set` treats `selfAtom` as a special case, but its type signature doesn't reflect this. One example is the implementation of `atomWithReducer`:

```typescript
export function atomWithReducer<Value, Action>(
  initialValue: Value,
  reducer: (value: Value, action: Action) => Value
) {
  const anAtom: any = atom(initialValue, (get, set, action: Action) =>
    set(anAtom, reducer(get(anAtom), action))
  )
  return anAtom
}
```

When we try to remove the `any`, typescript would complain at the call to `set` that "'Action' could be instantiated with an arbitrary type which could be unrelated to 'Value'". This is because `set` is expected to invoke write function, but when the first arg is the atom itself, it sets atom's value instead. So it would be nice to be able to distinguish the two and allow people to write something similar to `atomWithReducer` without having to use `any`.

Something like `set.self(newValue)` might be more desirable than having to rely on type assertion. It does the same thing as what `set(thisAtom, newValue)` does now, but has correct types. Also it's more explicit about the fact that we're setting the value instead of invoking the write function.

Although `set.recurse` might not be necessary, it's logical to have it once we have `set.self`, because this is what `set(thisAtom)` suggests it does in its original signature.

My suggestion is to extend `Setter` from:

```typescript
type Setter = <Value, Args extends unknown[], Result>(atom: WritableAtom<Value, Args, Result>, ...args: Args) => Result;
```
to something like:
```typescript
type Setter<SelfAtom> = {
  <Value, Args extends unknown[], Result>(atom: WritableAtom<Value, Args, Result>, ...args: Args): Result;
  // sets value directly.
  self(newValue: /* SelfValue */): void;
  // recursively call 
  recurse(...args: /* SelfArgs */): /* SelfResult */;
};
```

Because it's only an extra property on the setter function, it shouldn't break any existing code.

If you like this approach, I'd be happy to try to create a PR for it.

**Reply by n-p-e** [reply]
My own opinion is that `set` should do what its type signature suggests it does: always invoke the write function on the atom. Its special case seems a little bit confusing. But since it already has the special case, it might be better to discourage the use of `set(thisAtom, ...)` and instead make people choose `set.self` and `set.recurse` explicitly.

**Reply by dai-shi** [reply]
That I agree, I would like to make it to be `never`. We would like to let people use multiple atoms in this use case.

**Comment by dakom** [comment]
how about `setSelf` ?

This seems to work (though with the Suspense it only shows "loading"):

```
import { Suspense } from "react";
import { atom, useAtomValue } from "jotai";
import "./styles.css";

const timerAtom = atom(0);

const loopAtom = atom<Promise<number>>(
  async (get, { setSelf }) => {
    const prevTime = get(timerAtom);
    const currTime = await new Promise(requestAnimationFrame);

    const deltaTime = currTime - prevTime;

    setSelf();
    return currTime;
  },
  (get, set) => {
    set(timerAtom, performance.now());
  }
);

const valueAtom = atom(async (get) => {
  const value = await get(loopAtom);
  console.log(value);
  return value;
});

function Value() {
  const value = useAtomValue(valueAtom);
  return <div>{value}</div>;
}

const App = () => (
  <Suspense fallback="Loading...">
    <div className="App">
      <h1>Hello Jotai</h1>
      <h2>Enjoy coding!</h2>
      <Value />
    </div>
  </Suspense>
);

export default App;

```

**Reply by dai-shi** [reply]
`setSelf` is an escape hatch to allow a hack in `read`. The read function is declarative and pure, so it's not generally recommended to use `setSelf`.

The OP case is about the write function, which is imperative and can be impure. If possible, we should do things in `write`.

**Comment by dakom** [comment]
I don't see a way to do recursion with only atoms that way, since something has to be pushed for it to trigger the writes.

So the idea here if we want it to only trigger an update when the atom changed would be to do it in useEffect? e.g.

```
const tickAtom = atom<number>(0)

const valueAtom = atom<Promise<number>>(async (get) => {
  get(tickAtom)
  await new Promise((r) => setTimeout(r, 1000))
  return performance.now()
})

function Value() {
  const value = useAtomValue(valueAtom)
  const setTick = useSetAtom(tickAtom)

  useEffect(() => {
    setTick(performance.now())
  }, [setTick, value])

  return <div>{value}</div>
}
```

**Comment by t6adev** [comment]
Hi, I hope it can help.
https://codesandbox.io/s/timer-using-jotai-atomwithobservable-with-rxjs-5wh9hl?file=/src/App.tsx

**Reply by t6adev** [reply]
https://github.com/pmndrs/jotai/assets/6883692/83923c24-8956-46fc-a16f-5f702224f484
